README.md
# LoggerMulticastBackend

[![Hex version](https://img.shields.io/hexpm/v/logger_multicast_backend.svg "Hex version")](https://hex.pm/packages/logger_multicast_backend)

A backend for `Logger` that delivers messages over multicast UDP.

Designed for headless embedded applications, it allows watching the log over the local network.

## Usage

The package can be installed by adding `logger_multicast_backend` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:logger_multicast_backend, "~> 0.2"}]
end
```

## Easy Defaults

In your logger config, simply do something like this:

```elixir
config :logger,
        backends: [ :console, LoggerMulticastBackend ]
```

or, at runtime, you can add this to your current config...

```elixir
Logger.add_backend LoggerMulticastBackend
```

Now, you'll have logging messages sent out on the default target multicast address, which is 224.0.0.224:9999.

To capture log messages on systems with [socat](http://www.dest-unreach.org/socat/), run:

```
$ socat - udp-recv:9999,ip-add-membership=224.0.0.224:eth0
```

or use [cell](https://github.com/nerves-project/cell-tool):

```
$ cell watch
```

## Custom Configuration

Don't like the default multicast target or format? Change it by including options like this:

```elixir
  config :logger, :logger_multicast_backend,
    target: {{224,1,22,223}, 4252},
    format: "$time $metadata[$level] $message\\n",
    level:  :info
```

The full range of custom configuration options in the tuple are as follows:

- __target__ - a tuple of the target unicast or multicast address and port, like {{241,0,0,3}, 52209}

- __level__ - the level to be logged by this backend. Note that messages are first filtered by the general level configuration in :logger

- __format__ - the format message used to print logs.
Defaults to: ``"$time $metadata[$level] $levelpad$message\n"``

- __metadata__ - the metadata to be printed by $metadata.
Defaults to an empty list (no metadata)


CHANGELOG.md
## v0.2.2

- fixed configuration API (thanks @fhunleth) and updated README

## v0.2.1

- cleaned up build for elixir 1.4
- hex package improvements

## v0.2.0

Major re-architecture

- Now handles port coming/going gracefully
- Caches all data sent when port is not open, and re-sends it when it is
- Keeps up to the last 1024 lines of log for a port that is closed
- Throttles UDP traffic to 10ms packet pacing
- Tolerates IP address changes properly
- Removed blocking send (speeds up logger overall)

## v0.1.0

- Initial module, lots of bugs


test/logger_multicast_backend_test.exs
defmodule LoggerMulticastBackendTest do

  @moduledoc """
  Extremely braindead test for now.
  """
  
  use ExUnit.Case
  require Logger

  test "logging doesn't crash" do
    Logger.add_backend LoggerMulticastBackend
    Logger.debug "this should write to the multicast socket!  make sure it does"
    Logger.debug "this should write more.  make sure it does"
    :timer.sleep 1000
  end
  
end



mix.exs
defmodule LoggerMulticastBackend.Mixfile do

  use Mix.Project

  @version "0.2.2"
  @github_url "https://github.com/ghitchens/logger_multicast_backend"

  def project do
    [app: :logger_multicast_backend,
     version: @version,
     elixir: "~> 1.4",
     description: description(),
     package: package(),
     deps: deps(),
     # info for docs
     source_url: @github_url,
     docs: [main: "LoggerMulticastBackend",
            extras: ["README.md", "CHANGELOG.md"]]]
  end

  defp description do
    "A Logger backend that logs via Multicast UDP"
  end

  defp package do
    [maintainers: ["Garth Hitchens"],
     licenses: ["MIT"],
     links: %{"GitHub" => @github_url}]
  end

  def application do
    [applications: [:logger]]
  end

  defp deps do
    [{:ex_doc, "~> 0.15", only: :dev}]
  end
end


lib/logger_multicast_sender.ex
defmodule LoggerMulticastSender do

  @moduledoc false

  use GenServer
  require Logger

  @socket_retry_time  1000      # milliseconds between port open attempts
  @packet_pacer_time  10        # milliseconds between multicasts
  @queue_size         1024      # maximum lines to save in the queue

  @socket_opts [:binary, {:broadcast, true}, {:active, false}, {:reuseaddr, true} ]

  def init(target) do
    Logger.debug "started multicast logging sender on target: #{inspect target}"
    state = %{target: target, queue: [], socket: nil}
    {:ok, state}
  end

  def handle_cast({:add_entry, entry}, state) do
    queue = state.queue ++ [entry]
    if (length(queue) > @queue_size) do
      [_ | queue_tail] = queue
      {:noreply, %{ state | queue: queue_tail}, @packet_pacer_time}
    else
      {:noreply, %{ state | queue: queue}, @packet_pacer_time}
    end
  end

  # "handle a timer with no current socket - attempt to open it"
  def handle_info(:timeout, %{socket: nil} = state) do
    case :gen_udp.open(0, @socket_opts) do
      {:ok, socket} ->
        {:noreply, %{ state | socket: socket }, @packet_pacer_time}
      _ ->
        {:noreply, state, @socket_retry_time}
    end
  end

  # "handle a timer with a valid socket - attempt to write to it"
  def handle_info(:timeout, %{queue: queue, socket: socket, target: {addr, port}} = state) when (length(queue) > 0) do
    [first | rest] = queue
    case :gen_udp.send(socket, addr, port, first) do
      :ok ->
        {:noreply, %{state | queue: rest}, @packet_pacer_time}
      _ ->
        :gen_udp.close socket
        {:noreply, %{state | socket: nil}, @socket_retry_time}
    end
  end

  def handle_info(:timeout, state) do
    {:noreply, state}
  end

end


lib/logger_multicast_backend.ex
defmodule LoggerMulticastBackend do

  @moduledoc """
  A backend for `Logger` that delivers messages over multicast UDP.

  Designed for headless embedded applications, it allows watching the log over the local network.

  ## Easy Defaults

  In your logger config, simply do something like this:

  ```elixir
  config :logger,
          backends: [ :console, LoggerMulticastBackend ]
  ```

  or, at runtime, you can add this to your current config...

  ```elixir
  Logger.add_backend LoggerMulticastBackend
  ```

  Now, you'll have logging messages sent out on the default target multicast address, which is 224.0.0.224:9999.

  ## Custom Configuration

Don't like the default multicast target or format? Change it by including options like this:

  ```elixir
  config :logger, :logger_multicast_backend,
      target: {{224,1,22,223}, 4252},
      format: "$time $metadata[$level] $message\\n",
      level:  :info
  ```

  The full range of custom configuration options in the tuple are as follows:

  - __target__ - a tuple of the target unicast or multicast address and port, like {{241,0,0,3}, 52209}

  - __level__ - the level to be logged by this backend. Note that messages are first filtered by the general level configuration in :logger

  - __format__ - the format message used to print logs.
  Defaults to: ``"$time $metadata[$level] $levelpad$message\n"``

  - __metadata__ - the metadata to be printed by $metadata.
  Defaults to an empty list (no metadata)

  """

  use GenEvent
  require Logger

  # @type level     :: Logger.level
  # @type format    :: String.t
  # @type metadata  :: [atom]

  @default_target {{224,0,0,224}, 9999}
  @default_format "$time $metadata[$level] $message\n"

  @doc false
  def init({__MODULE__, opts}) when is_list(opts) do
    env = Application.get_env(:logger, :logger_multicast_backend, [])
    opts = Keyword.merge(env, opts)
    Application.put_env(:logger, :logger_multicast_backend, opts)

    level  = Keyword.get(opts, :level)
    metadata = Keyword.get(opts, :metadata, [])
    format_opts = Keyword.get(opts, :format, @default_format)
    format = Logger.Formatter.compile(format_opts)
    target = Keyword.get(opts, :target, @default_target)

    Logger.debug "starting multicast backend on target #{inspect target}"
    {:ok, sender} = GenServer.start_link(LoggerMulticastSender, target)

    state = %{sender: sender, level: level, format: format, metadata: metadata}
    {:ok, state}
  end

  def init(__MODULE__), do: init({__MODULE__, []})

  @doc false
  def handle_event({level, _gl, {Logger, message, timestamp, metadata}}, %{level: min_level} = state) do
    if is_nil(min_level) or Logger.compare_levels(level, min_level) != :lt do
      entry = format_event(level, message, timestamp, metadata, state)
      :ok = GenServer.cast(state.sender, {:add_entry, entry})
    end
    {:ok, state}
  end

  defp format_event(level, msg, ts, md, %{format: format, metadata: metadata}) do
    Logger.Formatter.format(format, level, msg, ts, Keyword.take(md, metadata))
  end

end
